adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/JwtAuthHandler.java

```diff
@@ -0,0 +1,182 @@
+
+    private final PreCredentialsValidationHandler<HttpContext> preCredentialsValidationHandler;
+
+    /**
+     * Creates a new handler for an authentication provider and a Tenant service client.
+     *
+     * @param authProvider The authentication provider to use for verifying the device identity.
+     * @param realm The realm name.
+     * @throws NullPointerException if client auth is {@code null}.
+     */
+    public JwtAuthHandler(final DeviceCredentialsAuthProvider<JwtCredentials> authProvider, final String realm) {
+        this(authProvider, realm, null);
+    }
+
+    /**
+     * Creates a new handler for an authentication provider and a Tenant service client.
+     *
+     * @param authProvider The authentication provider to use for verifying the device identity.
+     * @param realm The realm name.
+     * @param preCredentialsValidationHandler An optional handler to invoke after the credentials got determined and
+     *            before they get validated. Can be used to perform checks using the credentials and tenant information
+     *            before the potentially expensive credentials validation is done. A failed future returned by the
+     *            handler will fail the corresponding authentication attempt.
+     * @throws NullPointerException if client auth is {@code null}.
+     */
+    public JwtAuthHandler(final DeviceCredentialsAuthProvider<JwtCredentials> authProvider,
+            final String realm,
+            final PreCredentialsValidationHandler<HttpContext> preCredentialsValidationHandler) {
+        super(authProvider, Type.BEARER, realm);
+        this.preCredentialsValidationHandler = preCredentialsValidationHandler;
+    }
+
+    @Override
+    public void authenticate(final RoutingContext context, final Handler<AsyncResult<User>> handler) {
+
+        parseAuthorization(context, parseAuthorization -> {
+            if (parseAuthorization.failed()) {
+                handler.handle(Future.failedFuture(parseAuthorization.cause()));
+                return;
+            }
+
+            final String token = parseAuthorization.result();
+            int segments = 0;
+            for (int i = 0; i < token.length(); i++) {
+                final char c = token.charAt(i);
+                if (c == '.') {
+                    if (++segments == 3) {
+                        handler.handle(Future.failedFuture(new HttpException(HttpURLConnection.HTTP_BAD_REQUEST, "Too many segments in token")));
+                        return;
+                    }
+                    continue;
+                }
+                if (Character.isLetterOrDigit(c) || c == '-' || c == '_') {
+                    continue;
+                }
+                // invalid character
+                handler.handle(Future.failedFuture(new HttpException(HttpURLConnection.HTTP_BAD_REQUEST, "Invalid character in token: " + (int) c)));
+                return;
+            }
```

Comment:
isn't all of this already being taken care of by the DefaultJwsValidator.getJwtClaims() method?

Reply:
good catch, I have removed it.